src/settings/toml/migrations.rs

```diff
@@ -1,44 +1,175 @@
+use std::collections::HashSet;
+
 use serde::{Deserialize, Serialize};
 
-#[derive(Clone, Debug, Default, Deserialize, PartialEq, Serialize)]
-pub struct Migrations {
-    pub migrations: Vec<MigrationConfig>,
+#[derive(Clone, Debug, Deserialize, PartialEq, Serialize)]
+pub enum MigrationTag {
+    HasTag(String),
+    NoTag,
+    NoScript,
+    Unknown,
+}
+
+#[derive(Clone, Debug, Deserialize, PartialEq, Serialize)]
+pub enum Migrations {
+    Adhoc {
+        // The actual existing migration tag for the script, filled in by
+        // the publish command before form building. Used to validate provided_old_tag
+        script_tag: MigrationTag,
+        // The expected migration tag for the script if provided on the command line using --old-tag.
+        provided_old_tag: Option<String>,
+        // The new migration tag for the script if provided on the command line using --new-tag.
+        new_tag: Option<String>,
+        // The actual migration, if any directives are provided
+        migration: Option<Migration>,
+    },
+    List {
+        // The actual existing migration tag for the script, filled in by
+        // the publish command before form building
+        script_tag: MigrationTag,
+        migrations: Vec<MigrationConfig>,
+    },
 }
 
 impl Migrations {
-    pub fn api_migration(&self) -> Result<ApiMigration, anyhow::Error> {
-        // TODO: make api call to get most recent tag, and coalesce tags afterwards.
-        // For now, migrations will only ever have a single adhoc migration in it.
-        match &self.migrations {
-            migrations if migrations.len() == 1 => Ok(ApiMigration {
-                old_tag: None,
+    fn validate(&self) -> Result<(), anyhow::Error> {
+        // validate migration tags
+        match &self {
+            Migrations::Adhoc {
+                script_tag: MigrationTag::HasTag(script_tag),
+                provided_old_tag: Some(provided_tag),
+                ..
+            } if provided_tag != script_tag => anyhow::bail!(
+                "The migration tag provided with your adhoc migration (\"{}\") does not match the script's current migration tag of \"{}\", \
```

Comment:
nit, but do users know what "adhoc" means in this context? I'd leave that word out.

Or if we leave it in, "ad hoc" is the much more common way of spelling it :)